// @babel/helper-create-class-features-plugin/lib/index.js
"use strict";

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "FEATURES", {
  enumerable: true,
  get: function get() {
    return _features.FEATURES;
  }
});
Object.defineProperty(exports, "buildCheckInRHS", {
  enumerable: true,
  get: function get() {
    return _fields.buildCheckInRHS;
  }
});
exports.createClassFeaturePlugin = createClassFeaturePlugin;
Object.defineProperty(exports, "enableFeature", {
  enumerable: true,
  get: function get() {
    return _features.enableFeature;
  }
});
Object.defineProperty(exports, "injectInitialization", {
  enumerable: true,
  get: function get() {
    return _misc.injectInitialization;
  }
});
var _core = require("@babel/core");
var _helperFunctionName = require("@babel/helper-function-name");
var _helperSplitExportDeclaration = require("@babel/helper-split-export-declaration");
var _decorators = require("./decorators.js");
var _semver = require("semver");
var _fields = require("./fields.js");
var _decorators2 = require("./decorators-2018-09.js");
var _misc = require("./misc.js");
var _features = require("./features.js");
var _typescript = require("./typescript.js");
var versionKey = "@babel/plugin-class-features/version";
function createClassFeaturePlugin(_ref2) {
  var name = _ref2.name,
    feature = _ref2.feature,
    loose = _ref2.loose,
    manipulateOptions = _ref2.manipulateOptions,
    api = _ref2.api,
    inherits = _ref2.inherits,
    decoratorVersion = _ref2.decoratorVersion;
  var _api$assumption;
  if (feature & _features.FEATURES.decorators) {
    {
      if (decoratorVersion === "2023-11" || decoratorVersion === "2023-05" || decoratorVersion === "2023-01" || decoratorVersion === "2022-03" || decoratorVersion === "2021-12") {
        return (0, _decorators["default"])(api, {
          loose: loose
        }, decoratorVersion, inherits);
      }
    }
  }
  {
    var _api;
    (_api = api) != null ? _api : api = {
      assumption: function assumption() {
        return void 0;
      }
    };
  }
  var setPublicClassFields = api.assumption("setPublicClassFields");
  var privateFieldsAsSymbols = api.assumption("privateFieldsAsSymbols");
  var privateFieldsAsProperties = api.assumption("privateFieldsAsProperties");
  var noUninitializedPrivateFieldAccess = (_api$assumption = api.assumption("noUninitializedPrivateFieldAccess")) != null ? _api$assumption : false;
  var constantSuper = api.assumption("constantSuper");
  var noDocumentAll = api.assumption("noDocumentAll");
  if (privateFieldsAsProperties && privateFieldsAsSymbols) {
    throw new Error("Cannot enable both the \"privateFieldsAsProperties\" and " + "\"privateFieldsAsSymbols\" assumptions as the same time.");
  }
  var privateFieldsAsSymbolsOrProperties = privateFieldsAsProperties || privateFieldsAsSymbols;
  if (loose === true) {
    var explicit = [];
    if (setPublicClassFields !== undefined) {
      explicit.push("\"setPublicClassFields\"");
    }
    if (privateFieldsAsProperties !== undefined) {
      explicit.push("\"privateFieldsAsProperties\"");
    }
    if (privateFieldsAsSymbols !== undefined) {
      explicit.push("\"privateFieldsAsSymbols\"");
    }
    if (explicit.length !== 0) {
      console.warn("[".concat(name, "]: You are using the \"loose: true\" option and you are") + " explicitly setting a value for the ".concat(explicit.join(" and ")) + " assumption".concat(explicit.length > 1 ? "s" : "", ". The \"loose\" option") + " can cause incompatibilities with the other class features" + " plugins, so it's recommended that you replace it with the" + " following top-level option:\n" + "\t\"assumptions\": {\n" + "\t\t\"setPublicClassFields\": true,\n" + "\t\t\"privateFieldsAsSymbols\": true\n" + "\t}");
    }
  }
  return {
    name: name,
    manipulateOptions: manipulateOptions,
    inherits: inherits,
    pre: function pre(file) {
      (0, _features.enableFeature)(file, feature, loose);
      {
        if (typeof file.get(versionKey) === "number") {
          file.set(versionKey, "7.24.1");
          return;
        }
      }
      if (!file.get(versionKey) || _semver.lt(file.get(versionKey), "7.24.1")) {
        file.set(versionKey, "7.24.1");
      }
    },
    visitor: {
      Class: function Class(path, _ref3) {
        var file = _ref3.file;
        var _ref;
        if (file.get(versionKey) !== "7.24.1") return;
        if (!(0, _features.shouldTransform)(path, file)) return;
        var pathIsClassDeclaration = path.isClassDeclaration();
        if (pathIsClassDeclaration) (0, _typescript.assertFieldTransformed)(path);
        var loose = (0, _features.isLoose)(file, feature);
        var constructor;
        var isDecorated = (0, _decorators2.hasDecorators)(path.node);
        var props = [];
        var elements = [];
        var computedPaths = [];
        var privateNames = new Set();
        var body = path.get("body");
        var _iterator = _createForOfIteratorHelper(body.get("body")),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var _path = _step.value;
            if ((_path.isClassProperty() || _path.isClassMethod()) && _path.node.computed) {
              computedPaths.push(_path);
            }
            if (_path.isPrivate()) {
              var _name = _path.node.key.id.name;
              var getName = "get ".concat(_name);
              var setName = "set ".concat(_name);
              if (_path.isClassPrivateMethod()) {
                if (_path.node.kind === "get") {
                  if (privateNames.has(getName) || privateNames.has(_name) && !privateNames.has(setName)) {
                    throw _path.buildCodeFrameError("Duplicate private field");
                  }
                  privateNames.add(getName).add(_name);
                } else if (_path.node.kind === "set") {
                  if (privateNames.has(setName) || privateNames.has(_name) && !privateNames.has(getName)) {
                    throw _path.buildCodeFrameError("Duplicate private field");
                  }
                  privateNames.add(setName).add(_name);
                }
              } else {
                if (privateNames.has(_name) && !privateNames.has(getName) && !privateNames.has(setName) || privateNames.has(_name) && (privateNames.has(getName) || privateNames.has(setName))) {
                  throw _path.buildCodeFrameError("Duplicate private field");
                }
                privateNames.add(_name);
              }
            }
            if (_path.isClassMethod({
              kind: "constructor"
            })) {
              constructor = _path;
            } else {
              elements.push(_path);
              if (_path.isProperty() || _path.isPrivate() || _path.isStaticBlock != null && _path.isStaticBlock()) {
                props.push(_path);
              }
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
        {
          if (!props.length && !isDecorated) return;
        }
        var innerBinding = path.node.id;
        var ref;
        if (!innerBinding || !pathIsClassDeclaration) {
          (0, _helperFunctionName["default"])(path);
          ref = path.scope.generateUidIdentifier((innerBinding == null ? void 0 : innerBinding.name) || "Class");
        }
        var classRefForDefine = (_ref = ref) != null ? _ref : _core.types.cloneNode(innerBinding);
        var privateNamesMap = (0, _fields.buildPrivateNamesMap)(classRefForDefine.name, privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose, props, file);
        var privateNamesNodes = (0, _fields.buildPrivateNamesNodes)(privateNamesMap, privateFieldsAsProperties != null ? privateFieldsAsProperties : loose, privateFieldsAsSymbols != null ? privateFieldsAsSymbols : false, file);
        (0, _fields.transformPrivateNamesUsage)(classRefForDefine, path, privateNamesMap, {
          privateFieldsAsProperties: privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose,
          noUninitializedPrivateFieldAccess: noUninitializedPrivateFieldAccess,
          noDocumentAll: noDocumentAll,
          innerBinding: innerBinding
        }, file);
        var keysNodes, staticNodes, instanceNodes, lastInstanceNodeReturnsThis, pureStaticNodes, classBindingNode, wrapClass;
        {
          if (isDecorated) {
            staticNodes = pureStaticNodes = keysNodes = [];
            var _ref4 = (0, _decorators2.buildDecoratedClass)(classRefForDefine, path, elements, file);
            instanceNodes = _ref4.instanceNodes;
            wrapClass = _ref4.wrapClass;
          } else {
            keysNodes = (0, _misc.extractComputedKeys)(path, computedPaths, file);
            var _ref5 = (0, _fields.buildFieldsInitNodes)(ref, path.node.superClass, props, privateNamesMap, file, setPublicClassFields != null ? setPublicClassFields : loose, privateFieldsAsSymbolsOrProperties != null ? privateFieldsAsSymbolsOrProperties : loose, noUninitializedPrivateFieldAccess, constantSuper != null ? constantSuper : loose, innerBinding);
            staticNodes = _ref5.staticNodes;
            pureStaticNodes = _ref5.pureStaticNodes;
            instanceNodes = _ref5.instanceNodes;
            lastInstanceNodeReturnsThis = _ref5.lastInstanceNodeReturnsThis;
            classBindingNode = _ref5.classBindingNode;
            wrapClass = _ref5.wrapClass;
          }
        }
        if (instanceNodes.length > 0) {
          (0, _misc.injectInitialization)(path, constructor, instanceNodes, function (referenceVisitor, state) {
            {
              if (isDecorated) return;
            }
            var _iterator2 = _createForOfIteratorHelper(props),
              _step2;
            try {
              for (_iterator2.s(); !(_step2 = _iterator2.n()).done;) {
                var prop = _step2.value;
                if (_core.types.isStaticBlock != null && _core.types.isStaticBlock(prop.node) || prop.node["static"]) continue;
                prop.traverse(referenceVisitor, state);
              }
            } catch (err) {
              _iterator2.e(err);
            } finally {
              _iterator2.f();
            }
          }, lastInstanceNodeReturnsThis);
        }
        var wrappedPath = wrapClass(path);
        wrappedPath.insertBefore([].concat(_toConsumableArray(privateNamesNodes), _toConsumableArray(keysNodes)));
        if (staticNodes.length > 0) {
          wrappedPath.insertAfter(staticNodes);
        }
        if (pureStaticNodes.length > 0) {
          wrappedPath.find(function (parent) {
            return parent.isStatement() || parent.isDeclaration();
          }).insertAfter(pureStaticNodes);
        }
        if (classBindingNode != null && pathIsClassDeclaration) {
          wrappedPath.insertAfter(classBindingNode);
        }
      },
      ExportDefaultDeclaration: function ExportDefaultDeclaration(path, _ref6) {
        var file = _ref6.file;
        {
          if (file.get(versionKey) !== "7.24.1") return;
          var decl = path.get("declaration");
          if (decl.isClassDeclaration() && (0, _decorators2.hasDecorators)(decl.node)) {
            if (decl.node.id) {
              (0, _helperSplitExportDeclaration["default"])(path);
            } else {
              decl.node.type = "ClassExpression";
            }
          }
        }
      }
    }
  };
}
