// @babel/helper-define-polyfill-provider/lib/imports-injector.js
"use strict";

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArrayLimit(r, l) { var t = null == r ? null : "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (null != t) { var e, n, i, u, a = [], f = !0, o = !1; try { if (i = (t = t.call(r)).next, 0 === l) { if (Object(t) !== t) return; f = !1; } else for (; !(f = (e = i.call(t)).done) && (a.push(e.value), a.length !== l); f = !0); } catch (r) { o = !0, n = r; } finally { try { if (!f && null != t["return"] && (u = t["return"](), Object(u) !== u)) return; } finally { if (o) throw n; } } return a; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _createForOfIteratorHelper(o, allowArrayLike) { var it = typeof Symbol !== "undefined" && o[Symbol.iterator] || o["@@iterator"]; if (!it) { if (Array.isArray(o) || (it = _unsupportedIterableToArray(o)) || allowArrayLike && o && typeof o.length === "number") { if (it) o = it; var i = 0; var F = function F() {}; return { s: F, n: function n() { if (i >= o.length) return { done: true }; return { done: false, value: o[i++] }; }, e: function e(_e) { throw _e; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var normalCompletion = true, didErr = false, err; return { s: function s() { it = it.call(o); }, n: function n() { var step = it.next(); normalCompletion = step.done; return step; }, e: function e(_e2) { didErr = true; err = _e2; }, f: function f() { try { if (!normalCompletion && it["return"] != null) it["return"](); } finally { if (didErr) throw err; } } }; }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, _toPropertyKey(descriptor.key), descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _toPropertyKey(t) { var i = _toPrimitive(t, "string"); return "symbol" == _typeof(i) ? i : i + ""; }
function _toPrimitive(t, r) { if ("object" != _typeof(t) || !t) return t; var e = t[Symbol.toPrimitive]; if (void 0 !== e) { var i = e.call(t, r || "default"); if ("object" != _typeof(i)) return i; throw new TypeError("@@toPrimitive must return a primitive value."); } return ("string" === r ? String : Number)(t); }
function _typeof(o) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (o) { return typeof o; } : function (o) { return o && "function" == typeof Symbol && o.constructor === Symbol && o !== Symbol.prototype ? "symbol" : typeof o; }, _typeof(o); }
exports.__esModule = true;
exports["default"] = void 0;
var _babel = _interopRequireWildcard(require("@babel/core"));
function _getRequireWildcardCache(nodeInterop) {
  if (typeof WeakMap !== "function") return null;
  var cacheBabelInterop = new WeakMap();
  var cacheNodeInterop = new WeakMap();
  return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) {
    return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
  })(nodeInterop);
}
function _interopRequireWildcard(obj, nodeInterop) {
  if (!nodeInterop && obj && obj.__esModule) {
    return obj;
  }
  if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") {
    return {
      "default": obj
    };
  }
  var cache = _getRequireWildcardCache(nodeInterop);
  if (cache && cache.has(obj)) {
    return cache.get(obj);
  }
  var newObj = {};
  var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
  for (var key in obj) {
    if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
      var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
      if (desc && (desc.get || desc.set)) {
        Object.defineProperty(newObj, key, desc);
      } else {
        newObj[key] = obj[key];
      }
    }
  }
  newObj["default"] = obj;
  if (cache) {
    cache.set(obj, newObj);
  }
  return newObj;
}
var _ref = _babel["default"] || _babel,
  t = _ref.types;
var ImportsCachedInjector = /*#__PURE__*/function () {
  function ImportsCachedInjector(resolver, getPreferredIndex) {
    _classCallCheck(this, ImportsCachedInjector);
    this._imports = new WeakMap();
    this._anonymousImports = new WeakMap();
    this._lastImports = new WeakMap();
    this._resolver = resolver;
    this._getPreferredIndex = getPreferredIndex;
  }
  return _createClass(ImportsCachedInjector, [{
    key: "storeAnonymous",
    value: function storeAnonymous(programPath, url, moduleName, getVal) {
      var key = this._normalizeKey(programPath, url);
      var imports = this._ensure(this._anonymousImports, programPath, Set);
      if (imports.has(key)) return;
      var node = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)));
      imports.add(key);
      this._injectImport(programPath, node, moduleName);
    }
  }, {
    key: "storeNamed",
    value: function storeNamed(programPath, url, name, moduleName, getVal) {
      var key = this._normalizeKey(programPath, url, name);
      var imports = this._ensure(this._imports, programPath, Map);
      if (!imports.has(key)) {
        var _getVal = getVal(programPath.node.sourceType === "script", t.stringLiteral(this._resolver(url)), t.identifier(name)),
          node = _getVal.node,
          id = _getVal.name;
        imports.set(key, id);
        this._injectImport(programPath, node, moduleName);
      }
      return t.identifier(imports.get(key));
    }
  }, {
    key: "_injectImport",
    value: function _injectImport(programPath, node, moduleName) {
      var _this$_lastImports$ge;
      var newIndex = this._getPreferredIndex(moduleName);
      var lastImports = (_this$_lastImports$ge = this._lastImports.get(programPath)) != null ? _this$_lastImports$ge : [];
      var isPathStillValid = function isPathStillValid(path) {
        return path.node &&
        // Sometimes the AST is modified and the "last import"
        // we have has been replaced
        path.parent === programPath.node && path.container === programPath.node.body;
      };
      var last;
      if (newIndex === Infinity) {
        // Fast path: we can always just insert at the end if newIndex is `Infinity`
        if (lastImports.length > 0) {
          last = lastImports[lastImports.length - 1].path;
          if (!isPathStillValid(last)) last = undefined;
        }
      } else {
        var _iterator = _createForOfIteratorHelper(lastImports.entries()),
          _step;
        try {
          for (_iterator.s(); !(_step = _iterator.n()).done;) {
            var _step$value = _slicedToArray(_step.value, 2),
              i = _step$value[0],
              data = _step$value[1];
            var path = data.path,
              index = data.index;
            if (isPathStillValid(path)) {
              if (newIndex < index) {
                var _path$insertBefore = path.insertBefore(node),
                  _path$insertBefore2 = _slicedToArray(_path$insertBefore, 1),
                  newPath = _path$insertBefore2[0];
                lastImports.splice(i, 0, {
                  path: newPath,
                  index: newIndex
                });
                return;
              }
              last = path;
            }
          }
        } catch (err) {
          _iterator.e(err);
        } finally {
          _iterator.f();
        }
      }
      if (last) {
        var _last$insertAfter = last.insertAfter(node),
          _last$insertAfter2 = _slicedToArray(_last$insertAfter, 1),
          _newPath = _last$insertAfter2[0];
        lastImports.push({
          path: _newPath,
          index: newIndex
        });
      } else {
        var _programPath$unshiftC = programPath.unshiftContainer("body", node),
          _programPath$unshiftC2 = _slicedToArray(_programPath$unshiftC, 1),
          _newPath2 = _programPath$unshiftC2[0];
        this._lastImports.set(programPath, [{
          path: _newPath2,
          index: newIndex
        }]);
      }
    }
  }, {
    key: "_ensure",
    value: function _ensure(map, programPath, Collection) {
      var collection = map.get(programPath);
      if (!collection) {
        collection = new Collection();
        map.set(programPath, collection);
      }
      return collection;
    }
  }, {
    key: "_normalizeKey",
    value: function _normalizeKey(programPath, url) {
      var name = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : "";
      var sourceType = programPath.node.sourceType;

      // If we rely on the imported binding (the "name" parameter), we also need to cache
      // based on the sourceType. This is because the module transforms change the names
      // of the import variables.
      return "".concat(name && sourceType, "::").concat(url, "::").concat(name);
    }
  }]);
}();
exports["default"] = ImportsCachedInjector;
